### ISSUE TITLE
feat(ingest/elasticsearch) Add support for api_key authentication

### ISSUE THREAD

**Issue by th0ger** [datahub-project/datahub#11473]
**Is your feature request related to a problem? Please describe.**

Add the ability to use API key authentication for [Elasticsearch sources](https://datahubproject.io/docs/generated/ingestion/sources/elasticsearch/), as an alternative to `username`/`password` and certificate-based auth.
Proposing to add optional `api_key` (string | tuple) config parameter.

**Describe the solution you'd like**

- Elasticsearch accepts API key authentication via [http headers in the format](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/security-api-create-api-key.html#security-api-create-api-key-example): `"Authorization: ApiKey [ENCODED]"`

  The `[ENCODED]` key is provided by Elasticsearch during API key creation (internally it's a a base64 combination of `[ID]: [API_KEY]`).

- The Datahub Elasticsearch Source is implemented via [elasticsearch-py 7.x](https://elasticsearch-py.readthedocs.io/en/7.x/api.html#elasticsearch) (currently pinned version `elasticsearch==7.13.4`). This passes any auth kwargs via [Transport](https://elasticsearch-py.readthedocs.io/en/7.x/connection.html#elasticsearch.Transport) onto the default connection class [Urllib3HttpConnection ](https://elasticsearch-py.readthedocs.io/en/7.x/connection.html#urllib3httpconnection-default-connection-class) that accepts:
  *"**api_key** – optional API Key authentication as either base64 encoded string or a tuple."*

  Note that [elasticsearch-py latest 8.15.1](https://elasticsearch-py.readthedocs.io/en/latest/api/elasticsearch.html) takes the `api_key (str | Tuple[str, str] | None)` directly, which is compatible.
  
  I checked in the source ([_get_api_key_header_val](https://github.com/elastic/elasticsearch-py/blob/v7.13.4/elasticsearch/connection/base.py#L337-L346)) what they actually mean by "string or tuple", and it corresponds to the web-docs above. I.e. `api_key` requires either str `[ENCODED]` OR tuple `([ID], [API_KEY])`.

I propose the following datahub config format:

- Field: `api_key` (str | array), optional.
- Description: "API Key authentication. Accepts either a list with id and api_key (UTF-8 representation), or a base64 encoded string of id and api_key combined by ':'. "

Since YAML doesn't accept tuples, we need to accept a list and convert it to a python tuple.
From the looks of jsonschema and pydantic it should be possible to have a mixed type. Not sure if the datahub web documentation likes mixed types (anyone?). If it turns out not possible, we can just do encoded str.

**Describe alternatives you've considered**
NONE

**Additional context**
Tests: Datahub does not currently have integration tests for elasticsearch, and e.g. basic http_auth is currently also not tested. Besides setting up elasticsearch with docker-compose, testing api_key requires one to first create the api_key before testing it.
Creating tests are a much larger task than the feat itself, so needs clarification of scope.

**Comment by th0ger** [comment]
@hsheth2 Please review above spec and provide feedback on the 1) mixed types and 2) scope of required tests.

I can make a PR asap with the feat itself, but need more feedback for the tests.

**Comment by th0ger** [comment]
I just tested the the support for pydantic mixed Union and Tuple types.
Union is already used by datahub in various places.

`api_key: Optional[Union[Tuple[str,str], str]]` is handled well by Pydantic (it generates a jsonschema with anyOf and a list of fixed length 2):

![image](https://github.com/user-attachments/assets/8d387d09-ad7b-483e-a859-a65275934ed5)

But the Docusaurus table view have issues handling the `Tuple` type:
![image](https://github.com/user-attachments/assets/02a84654-0954-4a58-a975-bb1d966656b8)

**Comment by th0ger** [comment]
The reason for the missing table in the generated docs are caused by incompatible jsonschema versions.

The `./scripts/docgen.sh` is validating against `https://json-schema.org/draft/2020-12/schema` and fails validation.

The full generated jsonschema is pasted at the bottom of this post. You can put it into e.g. [JSON Schema Validator](https://www.jsonschemavalidator.net) and find that it validates OK for all versions up to `https://json-schema.org/draft/2019-09/schema` but indeed fails `https://json-schema.org/draft/2020-12/schema`.

The key part that's failing `draft/2020-12` is the Tuple construct:

```json
{
"type": "object",
"properties": {
  "api_key": {
    "anyOf": [
      {
        "type": "array",
        "minItems": 2,
        "maxItems": 2,
        "items": [
          {
            "type": "string"
          },
          {
            "type": "string"
          }
        ]
      },
      {
        "type": "string"
      }
    ]
  }
}
}
```

Indeed, `draft/2020-12` introduced breaking [changes-to-items-and-additionalitems](https://json-schema.org/draft/2020-12/release-notes#changes-to-items-and-additionalitems):

> The keywords used for defining arrays and **tuples have been redesigned** [...].
> The `items` and `additionalItems` keywords have been replaced with `prefixItems` and `items` [...].
> Although the meaning of items has changed, the syntax for defining arrays remains the same. **Only the syntax for defining tuples** has changed. 

**In summary**: I think Datahub generates json schemas compatible with `draft/2019-09` or below (not sure which and how), but attempts validation against `draft/2020-12`. This bug surfaces the water when using advanced constructs like tuples.

**Comment by th0ger** [comment]
@hsheth2 Thanks for taking in this change 🥳
